### ISSUE TITLE
The way `default_field_values` deals with in-scope generic parameters is slightly unprincipled

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust#146496]
During wfcheck we currently `const_eval_poly` default field values unless its ascribed type references generic parameters. However, this doesn't take into account generic parameters referenced by the const expr (this matters because the final const value or its type doesn't necessarily need to reference them).

This is problematic as const-eval bails out early with `TooGeneric` if it (in some underspecified(?) value-based fashion) encounters a type or const param meaning it does not report an error! This leads to surprising (and imo unwanted) behavior. For example, under the current implementation downstream can break if an upstream const fn stops using a type or const param in its implementation (something you wouldn't think should affect callers I'd say):

```rs
//- upstream -//
pub const fn f<const N: usize>() {
    let _ = [0u8; N]; // <-- comment out this line to break downstream!
}

//- downstream -//
#![feature(default_field_values)]

struct Container<const N: usize> {
    field: () = {
        upstream::f::<N>();
        panic!();
    },
}
```

While const-evaluating the default value of `Container.field`, if the compiler encounters const param `N` "in a const value" (so e.g., `let _: [u8; N];` wouldn't count) it'll silently bail out, not const-panic and thus succeed the build, otherwise it'll reach & execute the `panic!();` and therefore fail the build *pre-mono*.

<details><summary>Alternative example involving a type param instead of a const param</summary>

```rs
//- upstream -//
pub trait Test { const TEST: bool; } // or in some other crate further upstream

pub const fn f<T: Test>() {
    _ = T::TEST; // <-- comment out this line to break downstream!
}

//- downstream -//
#![feature(default_field_values)]

struct Container<T: upstream::Test> {
    field: () = {
        upstream::f::<T>();
        panic!() 
    },
    _marker: std::marker::PhantomData<T>,
}
```

</details>

Of course, the *final value* of public const items, const fns and the like can hardly be considered a private implementation detail (and modifying it would require a major version bump strictly speaking). However, this is about the const expr, the unevaluated body, where I'd hope there to be a bit more wiggle room.

---

**Regarding solutions**, I see two obvious ones: Either

1. reject const exprs that reference generic params (that's the approach we take for enum discriminant values) or
2. don't evaluate any default field values (at the def-site[^1]) if the enclosing ADT has
   1. any generic parameters or alternatively
   2. any "generic params that require monomorphization" (i.e., type and const params but not lifetime ones)
      * that's the approach we currently take for free generic const items (feature: `generic_const_items`)
      * since const-eval works modulo non-late-bound regions (they are erased just like during codegen), it's safe to eval the expr even if it references lifetime parameters
      * if we end up following this approach, then we should probably also relax the rule "don't eval if the ascribed field type contains generic params" to "[…] type/const params" and eval the initializer in `struct X<'a> { x: &'a () = panic!() }`

[^1]: At most use-sites we would still always evaluate as per the usual rules.

**Comment by fmease** [comment]
cc @estebank

**Comment by fmease** [comment]
In my opinion this is a bug but it's possible that other project members disagree.

**Comment by khalidsaidi** [comment]
A2ABench has an accepted answer for this imported thread.
- Thread: https://a2abench-api.web.app/q/cmmpdq81h00hl6t5u4jvqu9zb
- Accepted at: 2026-03-13T22:14:51.857Z
- Accepted answer agent: `live-ans-bulk-1773439863`
- Answer preview: "External agent resolution:\n\n1. Reproduce.\n2. Isolate root cause.\n3. Patch smallest surface.\n4. Verify with one negative and one positive test."

**Comment by fmease** [comment]
> I'd still want obviously failing consts like those to emit an error on the definition, somehow.

Right, well. This would need some kind of custom const expr visitor which doesn't walk into non-local functions etc[^1], hmm. I can understand why you'd want `panic!()` to get evaluated even if type&const params are in scope as long as they're "not used in the expr" according to some more conservative metric.

I don't want to make the call here. I'd rather WG-const-eval, T-lang and/or T-types(?) do that.

I'm just saying that under GCI `const _<const N: usize>: () = panic!();` does not get evaluated to be consistent with `fn f<const N: usize>() { const { panic!(); } }` which doesn't lead to a post-mono error due to that in-scope const param. It's also a lot simpler to implement & forward compatible with potential laxer rules.

[^1]: Even then, it wouldn't solve the analogous problem with non-local macros (however that might negligible / could be exempt from stability guarantees).

**Comment by bstrie** [comment]
> I don't want to make the call here. I'd rather WG-const-eval, T-lang and/or T-types(?) do that.

Given that this issue appears to currently be the only listed blocker for default field values, would someone like to nominate this for discussion?

**Comment by oli-obk** [comment]
As discussed in https://rust-lang.zulipchat.com/#narrow/channel/326866-t-types.2Fnominated/topic/.23146496.3A.20The.20way.20.60default_field_values.60.20deals.20with.20in-scope.E2.80.A6 this is a bug and should be fixed.

**Comment by estebank** [comment]
I would appreciate if someone from @rust-lang/wg-const-eval could take the lead on this change. I would like to stabilize the feature before EoY and I think this is the only outstanding issue and I want to make sure that it is done in a way that is acceptable to the WG.

**Comment by RalfJung** [comment]
Cc @rust-lang/types

**Comment by fmease** [comment]
While I did get a confirmation from T-types in the linked Zulip topic that this was a bug, in my eyes I didn't get a confirmation that they were also fine with the new semantics I've proposed. I've therefore gone ahead & opened PR https://github.com/rust-lang/rust/pull/163182 with the semantics I desire.
